Typescript LMAO

qm3ster opened the pull request:
The only breaking change is `writeStrcut` typo (and the corresponding Rsp) being corrected to `writeStruct`.
Has been tested with the `zigbee2mqtt` stack.

That's also what I used to profile what of the `"enum"` package's functionality is needed, not just in own methods.
It wasn't much. That module is essentially being misused here, so I eliminated it. @splitice also did similarly in his forks, replacing it with his `"light-enum"`.
But I believe the "enum" API doesn't represent a useful abstraction here, so I plan to eliminate it entirely, in favor of direct lookups.
The API of the local "enum" replacement in `enum.ts` (which I want to propagate up the stack) represents this.

I also eliminated the other dependency, `"busyman"` - a lodash alternative, in favor of native ES5 methods.
There are 0 runtime dependencies now.

Finally, I added decent(?) JSDoc documentation of exported objects and functions.

Also, it's set up to release in the `@zigbee` npm namespace.

I need this merged and regarded as the status quo so that I can reasonably go up the stack with
1. `zcl-id` types being provided
2. the specific methods on the `zcl-id` "enums" to replace the overloaded `.get`

qm3ster committed:
Update to org

Rename to scoped package
Update links, badges
Don't use make
Update travis confignpm

qm3ster committed:
Rename writeStrcut to writeStruct

Typo, technically breaking change

qm3ster committed:
0.5.0-0

qm3ster committed:
Fix travis deploy

qm3ster committed:
0.5.0-1

qm3ster committed:
Package the right files

qm3ster committed:
0.5.0-2

qm3ster committed:
Update dependencies

qm3ster committed:
Initial Typescript conversion

qm3ster committed:
Update travis config

Enable packing typescript artifacts

qm3ster committed:
Initial TypeScript conversion

qm3ster committed:
Furter refactoring

Eliminated dependencies

qm3ster committed:
0.5.0-4

qm3ster committed:
Further refactors

Added `keys` and `values` Maps to Enum for direct lookup.
This enables clearer and more performant code and will make it easier to
see which lookup tables are used / can be eliminated in the whole stack

Added more JSDoc descriptions

Elminated `null` from `cluster_defs.json`
Now the properties are just optional instead

Double quotes (Prettier default)

qm3ster committed:
0.5.0-5

splitice commented:
So essentially you built enum into the package? Hmm, personally I like the abstraction of a seperate package from a testing point of view.

One of the main reasons we switched to light-enum was the memory usage, light-enum saved us approx 20% of the memory in use on a simple test application. It doesn't rebuild new objects as often for the inverse mappings (remember each lookup works in both directions by default) and makes use of those loaded from source.

qm3ster commented:
I currently made the typings not depend on the `.json`s at all, they look like
```ts
interface CommonJSON {
    profileId: Record<string, number>;
    foundation: Record<string, number>;
    dataType: Record<string, number>;
    status: Record<string, number>;
    clusterId: Record<string, number>;
    haDevId: Record<string, number>;
    [key: string]: Record<string, number>;
}
```
but this enables `@zigbee/zcl-id` to be consumed by tsc without `--resolveJsonModule` option.

Other choices are:

### 1. Making it depend on the `.json`s 
As said above, this requires the consumer to have `--resolveJsonModule` flag or `tsconfig.json` setting.
This gives us access to the keys, but not the values.
eg, the type of `profileId` goes
```ts
// from
declare const profileId: Record<string, number>
// to
declare const profileId: Record<"HA" | "BA" | "TS" | "HC" | "SE" | "RS" | "LL", number>
```
### 2. Converting the `.json`s to `.ts`s
This is accomplished by adding `export default ` to the beginning of the file and hitting "save".
This gets us the same result as №1 but without depending on JSON modules.
We can also drop the quotes around keys, and add comments. (someone might be into this)

### 3. Generating `.ts` from the `.json` before compilation
This allows us to do the unthinkable: 
```ts
// instead of
export default { /* json here */ }
// we can do
export default { /* json here */ } as { /* json here again */ }
// which gets our types all the way to 
declare const profileId: {
    HA: 260;
    BA: 261;
    TS: 263;
    HC: 264;
    SE: 265;
    RS: 266;
    LL: 49246;
};
```
Do we actually want this? Maybe we do, maybe we don't.
Such enums are sometimes used, and it allows switch exhaustiveness checks, as well as catching typos all the way up the stack - "What are you handling `writeStrcut` for? We can never decode such a command name.", all without having to exercise that in a test.
As with any "strict" setting, this requires more handling at the edges - "Is this a `FoundationCommandName` or any `string` that this function accepts?" but once inside it allows for safe code without guards, once we know we aren't dealing with an arbitrary string.
An alternative to this would be using objects instead of the strings, to have classes like `Cluster` with fields like `attributes`, but this is incompatible with the current API, and has (an absolutely negligible) runtime impact.

qm3ster commented:
@splitice I probably have a negative impact on memory at the moment, each `Enum` is 4 maps + and array. (2 maps and the array are references to the same n objects though, and the input object is dropped after constructor)
As I mentioned above the goal of this `Enum` implementation is not to be anything that will go into public use, but just a stepping stone used to replace what the packages higher in the stack call, until I can switch them over to more pointed lookups by knowing the input type at the call site.

qm3ster committed:
Format common.json

qm3ster committed:
Add otaStatus enum

These are the codes described in **11.13.2 OTA Cluster Status Codes** [on page 701](http://www.zigbee.org/~zigbeeor/wp-content/uploads/2014/10/07-5123-06-zigbee-cluster-library-specification.pdf).

Originally added by @splitice in https://github.com/HalleyAssist/zcl-id/commit/34baf8c439eb553276f72d9f45597f7519131df3

qm3ster committed:
Format enum.ts

qm3ster committed:
Revert "Add support for TRADFRI remote clusters. https://github.com/Koenkk/zigbee2mqtt/issues/102"

This reverts commit 9a27ea86063627c5e36f529d7f9113702a18662b.

We aren't sure what these commands are yet,
but they will be possible to add through the extend hook.

qm3ster committed:
Initial commit of Hacks.md

qm3ster committed:
Add `structElms` to Hack.md

splitice commented:
I'm not keen on the introduction of a typescript step.

An extends method was our original plan, however we had some difficulties making that happen at the time as such a patched fork was the method we went with.

qm3ster committed:
Typos

Breaking, typos in identifiers

qm3ster committed:
Add constructor

Old singleton import can be achieved with `require("zcl-id/legacy")`

qm3ster committed:
0.5.0-6

qm3ster committed:
Add converter to new format

The source of truth will now be the denormalized deduplicated format
This also includes the responsibilities of `zcl_meta` from `zcl-packet`
Also fixed some missing information discovered

qm3ster commented:
@splitice
>I'm not keen on the introduction of a typescript step.

Do you mean for the JSONs? Yeah, not doing that currently.
Will maybe end up generating a `.d.ts` file for better suggestions, but the `JSON` will remain unmodified in the repo.
However, I am really considering JSON5, because having the numbers as decimals instead of hex sucks. Other repos and the spec PDFs all use hex, so it's bothersome when searching.
Not sure what to do about that.

Additionally, in e25fd3a, I started working on a new source-of-truth format.
For `cluster_defs.json` it looks like this:
```json
"genIdentify": {
  "id": 3,
  "attrs": {
    "identifyTime": {
      "id": 0,
      "type": "uint16"
    },
    "identifyCommissionState": {
      "id": 1,
      "type": "unknown"
    }
  },
  "cmd": {
    "identify": {
      "id": 0,
      "params": [["identifytime", "uint16"]]
    },
    "identifyQuery": {
      "id": 1,
      "params": []
    },
    "ezModeInvoke": {
      "id": 2,
      "params": [["action", "uint8"]]
    },
    "updateCommissionState": {
      "id": 3,
      "params": [["action", "uint8"], ["commstatemask", "uint8"]]
    },
    "triggerEffect": {
      "id": 64,
      "params": [["effectid", "uint8"], ["effectvariant", "uint8"]]
    }
  },
  "cmdRsp": {
    "identifyQueryRsp": {
      "id": 0,
      "params": [["timeout", "uint16"]]
    }
  }
```
The `param` tuples were inspired by your fork, but I don't think having types as numbers is very readable. As you can see, the `zcl_meta.json` data is all merged into here as well, likewise for functional commands.
I got rid of `knownBufLen` because its current version is literally just
```js
params.reduce((s, [_, t]) => s + (t === "uint8") || (t === "uint16") * 2)
```
(I checked)
I also got rid of `dir`, because all `cmd` are `dir: 0` and all `cmdRsp` are `dir: 1`
(I checked)

I am keeping object style for `attrs` at this time, because there may be other properties we may need to consider.
And yes, I realize `[id:number, type:string, writeable:bool, optional: bool]` or even `[id:number, type:string, options?: {writeable?:bool, optional?: bool}]` is an option, just not sure yet.

qm3ster committed:
Write converted jsons with no spaces

Run through prettier instead

qm3ster committed:
Commit converted jsons

qm3ster committed:
Switch accessors to new jsons, naively

Made tests pass without module depending on old JSONs
Old jsons still referenced by tests

qm3ster committed:
Move back from scoped name

qm3ster committed:
Update dependencies

qm3ster committed:
Update dependencies

qm3ster committed:
Export the 3 new jsons in index

qm3ster committed:
Add prettier npm scripts

qm3ster committed:
Switch to Jest

qm3ster committed:
Move zclmeta.ts from zcl-packet

qm3ster committed:
Don't depend on old zcl_meta.json

qm3ster committed:
Remove "use strict" from ts file

qm3ster committed:
Don't delete converted clusterDefs

zclmeta depends on them for now

qm3ster committed:
Remove unused import

qm3ster committed:
Move knownBufLen calculation to zcl-packet

qm3ster committed:
Fix some names to match ZCL docs

qm3ster committed:
Correct some parameter types to be more specific

qm3ster committed:
Update `discoverRsp` params to new format

qm3ster committed:
Update foundation parameters

Also removed conversion script to prevent overwriting
